### PR TITLE
fix: add erc721 token revoke approve

### PR DESCRIPTION
**Pull request by qiweiii**
resolves https://github.com/Magickbase/godwoken_explorer/issues/1030
resolves https://github.com/Magickbase/godwoken-explorer-ui/issues/618
resolves https://github.com/Magickbase/godwoken-explorer-ui/issues/655

Resolve problems:
- added erc721 approve type revoke
- added cache in frontend, so even user change tabs in account page, the revoke button's loading state is not lost
- prevent auto connect when open approval tab

### CODE REVIEW

**Comment by qiweiii** [comment]
@Keith-CY can we use `token_approvals` query’s `udt.id` as `tokenId` in erc721’s `approve(zero address, tokenId)` method call?

**Comment by Keith-CY** [comment]
> @Keith-CY can we use `token_approvals` query’s `udt.id` as `tokenId` in erc721’s `approve(zero address, tokenId)` method call?

`udt.id` is `id` in backend while `tokenId` in `approve(zero address, tokenId)` is `id` in the contract on-chain so they are different.

`udt.id` would change if we start a new server but `tokenId` is immutable since it's written in the contract.

**Comment by qiweiii** [comment]
@Keith-CY so we should use `token_contract_address_hash` as `tokenId`?

**Comment by Keith-CY** [comment]
> @Keith-CY so we should use `token_contract_address_hash` as `tokenId`?

No, `token Id` could be fetched from `event Approval(spender, tokenId)` and it should be returned in the record from the backend(if not, please assign it to backend developers)

When I approve someone to spend my NFT, I call the `function approve(address to, uint256 tokenId)`(https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol#L112) method and it will emit a log `emit Approval(ERC721.ownerOf(tokenId), to, tokenId)`(https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol#L394)

It simply changes an internal map `mapping(uint256 => address) private _tokenApprovals;` in the contract (https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol#L36) so the approval could be located by `contract address -> token id -> spender`

The log will be detected and stored in the server as https://testnet.gwscan.com/tx/0xbd6b5bf98298664733e7ca1a18d6712e326948e9d0362a3a01994e1d45d70e0f?tab=logs, `tokenId` is there, and revocation is the reverse action so `contract address` and `token id` are both necessary

**Comment by zmcNotafraid** [comment]
> @Keith-CY can we use `token_approvals` query’s `udt.id` as `tokenId` in erc721’s `approve(zero address, tokenId)` method call?

token_approval's data column was tokenID

**Comment by qiweiii** [comment]
> > @Keith-CY can we use `token_approvals` query’s `udt.id` as `tokenId` in erc721’s `approve(zero address, tokenId)` method call?
> 
> 
> 
> token_approval's data attribute was tokenID

Great!

**Comment by Keith-CY** [comment]
conflicted

File: components/TokenApprovalList/RevokeButton.tsx [hunk 1]
```diff
@@ -62,6 +70,9 @@ const RevokeButton: React.FC<Props> = ({ setAlert, listItem, account, hideItem }
   const currentContract = mapEthTypeToABI(listItem)
   const [callWrite, setCallWrite] = useState(false)
   const [isPackaging, setIsPackaging] = useState(false)
+  const { transaction_hash, spender_address_hash, token_contract_address_hash, data } = listItem
+  const itemKey = transaction_hash + spender_address_hash + token_contract_address_hash + data + '-revokeTxn'
+  const [revokeTxnHash, setRevokeTxnHash] = useState(localStorage.getItem(itemKey) || '')
```
Comment (Keith-CY) [inline components/TokenApprovalList/RevokeButton.tsx:75]: What if multiple approvals are revoked?

Comment (qiweiii) [reply components/TokenApprovalList/RevokeButton.tsx:75]: this is inside component RevokeButton, so each button has its own state, did I miss anything?

Comment (Keith-CY) [reply components/TokenApprovalList/RevokeButton.tsx:75]: > this is inside component RevokeButton, so each button has its own state, did I miss anything?

It's me misunderstood it

**Comment by FrederLu** [comment]
image
@qiweiii 
Clicking the data undo button in the screenshot does not invoke MetaMask.

**Comment by qiweiii** [comment]
> @qiweiii Clicking the data undo button in the screenshot does not invoke MetaMask. Other data in the screenshot can be evoked.

Fixed, try again

**Review by FrederLu — APPROVED** [review]

**Pull request merged**